File: Clustering/Hierarchical/hierarchical_clustering.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Hierarchical Clustering Implementation
Updated for better portability
"""

# Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
import scipy.cluster.hierarchy as sch
from sklearn.cluster import AgglomerativeClustering

# Get the directory where the script is located
script_dir = os.path.dirname(os.path.abspath(__file__))
file_path = os.path.join(script_dir, "Mall_Customers.csv")  # Dataset file

# Checking if the dataset exists
if os.path.isfile(file_path):
    dataset = pd.read_csv(file_path)
    print("Data imported successfully!")  
else:
    raise FileNotFoundError(f"Error: File not found: {file_path}. Please make sure 'Mall_Customers.csv' is in the same directory.")

# Selecting relevant features for clustering
X = dataset.iloc[:, [3, 4]].values  # Using Annual Income and Spending Score

# Using Dendrogram to find the optimal number of clusters
plt.figure(figsize=(10, 5))
dendrogram = sch.dendrogram(sch.linkage(X, method='ward'))
plt.title('Dendrogram for Hierarchical Clustering')
plt.xlabel('Customers')
plt.ylabel('Euclidean Distances')
plt.grid(True)
plt.show()

# Applying Hierarchical Clustering with the optimal number of clusters (K=5)
hc = AgglomerativeClustering(n_clusters=5, affinity='euclidean', linkage='ward')
y_hc = hc.fit_predict(X)

# Visualizing the Clusters
plt.figure(figsize=(8, 6))
plt.scatter(X[y_hc == 0, 0], X[y_hc == 0, 1], s=100, c='red', label='Frugal Spenders')
plt.scatter(X[y_hc == 1, 0], X[y_hc == 1, 1], s=100, c='blue', label='Moderate Spenders')
plt.scatter(X[y_hc == 2, 0], X[y_hc == 2, 1], s=100, c='green', label='Big Spenders')
plt.scatter(X[y_hc == 3, 0], X[y_hc == 3, 1], s=100, c='cyan', label='Careless Spenders')
plt.scatter(X[y_hc == 4, 0], X[y_hc == 4, 1], s=100, c='magenta', label='Sensible Spenders')
plt.title('Hierarchical Clustering: Customer Segments')
plt.xlabel('Annual Income (k$)')
plt.ylabel('Spending Score (1-100)')
plt.legend()
plt.grid(True)
plt.show()

File: Clustering/Hierarchical/Hierarchical_Clustering.R
# Hierarchical Clustering

# Importing Dataset
dataset <- read.csv('mall.csv') 
X = dataset[4:5]  # Selecting Annual Income and Spending Score

# Compute the distance matrix only once
distance_matrix <- dist(X, method = 'euclidean')

# Using Dendrogram to find the optimal number of clusters
dendrogram = hclust(distance_matrix, method = 'ward.D')

# Enhanced visualization with color and titles
plot(dendrogram,
     main = "Dendrogram for Optimal Clusters",
     xlab = "Number of Customers", 
     ylab = "Euclidean Distance",
     sub = "",
     cex = 0.9,  # Adjust font size
     hang = -1,  # Align labels at the bottom
     col = "blue")  

# Applying Hierarchical Clustering
hc = hclust(distance_matrix, method = 'ward.D')
optimal_clusters = 5  # You can dynamically determine this from the dendrogram
y_hc = cutree(hc, optimal_clusters)

# Visualizing the clusters
library(cluster)
clusplot(X,
         y_hc,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 2,
         plotchar = FALSE,
         span = TRUE,
         main = "Clusters of Customers",
         xlab = "Annual Income",
         ylab = "Spending Score",
         col.p = y_hc,  # Different colors for different clusters
         cex = 0.9)  # Adjust point size
